### ISSUE TITLE
The file "'tfbackend'" could not be read.

### ISSUE THREAD

**Issue by Setthawut-Peace** [DevSecTop/TF-via-PR#347]
Hi, I got error `The file "'env-dev/dev.azurerm.tfbackend'" could not be read.`
My terraform backend and tfvar directory is ./env-dev/dev.azurerm.tfbackend and ./env-dev/dev.tfvars. 
when I try remove "working-directory: ${{ github.workspace }}" got other error and file not found. I have 2 have 2 submodule in my repo both is terraform template code.

this is my workflow file
```
---
name: Trigger on schedule (cron) event with fmt/validate checks to open an issue on configuration drift.

on:
  workflow_dispatch:

permissions:
  actions: read        # Required to identify workflow run.
  checks: write        # Required to add status summary.
  contents: read       # Required to checkout repository.
  issues: write        # Required to open issue.
  pull-requests: write # Required to add comment and label.
  id-token: write

env:
  TF_WORKSPACE: dev
  ARM_USE_MSI: true
  ARM_SUBSCRIPTION_ID: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
  ARM_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
  ARM_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
  ARM_SKIP_PROVIDER_REGISTRATION: true

jobs:
  Terraform_Init_And_Plan:
    name: "Terraform Plan"
    runs-on: ubuntu-latest
    timeout-minutes: 10
    environment: dev

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Setup TF
        uses: hashicorp/setup-terraform@v3

      - name: "Setup: Load environment variables from env file"
        run: cat env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.env | grep -v ^# >> $GITHUB_ENV
        continue-on-error: true

      - name: "Setup: GitHub ACCESS_TOKEN with global configuration"
        run: |
          git config --global url."https://tarathec:${{ secrets.REPO_ACCESS_TOKEN }}@github.com".insteadOf "https://github.com"

      - name: "Setup: Pull & Update Submodules"
        run: |
          git submodule update --init --recursive
          git submodule update --recursive --remote
          chmod 777 terraform-utilities/github-action/*.sh

      - name: "Setup: Terraform Variables"
        run: |
          echo "TF_VAR_use_oidc=true" >> $GITHUB_ENV
          echo "TF_VAR_oidc_request_token=$ACTIONS_ID_TOKEN_REQUEST_TOKEN" >> $GITHUB_ENV
          echo "TF_VAR_oidc_request_url=$ACTIONS_ID_TOKEN_REQUEST_URL" >> $GITHUB_ENV

      - name: "Setup: Log in with Azure"
        uses: Azure/login@v2.2.0
        with:
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}

      - name: "Misc: Run Azure CLI for add network rule"
        run: terraform-utilities/github-action/add_network_rule.sh

      - name: "Misc: Generates SAS Token"
        run: terraform-utilities/github-action/get_sas_azurerm.sh
      
      - name: "Setup: waiting for Storage NSG"
        id: init
        run: |
          sleep 30

      - name: Provision TF
        id: provision
        uses: devsectop/tf-via-pr@v12
        with:
          command: plan
          working-directory: ${{ github.workspace }}
          arg-backend-config: env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.azurerm.tfbackend
          arg-var-file: env-${{ env.TF_WORKSPACE }}/${TF_WORKSPACE}.tfvars
          arg-lock: false
          arg-refresh-only: true
          format: false
          validate: false

      - name: Open issue on drift
        if: steps.provision.outputs.exitcode != 0
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          diff: ${{ steps.provision.outputs.diff }}
          run: ${{ steps.provision.outputs.run-url }}
          result: ${{ steps.provision.outputs.result }}
          summary: ${{ steps.provision.outputs.summary }}
        run: |
          gh api /repos/{owner}/{repo}/issues \
            --method POST \
            --field title="Configuration drift detected" \
            --field body="[View log.]($run)
          <details><summary>Diff of changes.</summary>

          \`\`\`diff
          $diff
          \`\`\`
          </details>
          <details><summary>$summary</summary>

          \`\`\`hcl
          $result
          \`\`\`
          </details>"
```

**Comment by rdhar** [comment]
Seems like a typo on the following line:

```diff
- arg-var-file: env-${{ env.TF_WORKSPACE }}/${TF_WORKSPACE}.tfvars
+ arg-var-file: env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.tfvars
```

And I don't think `working-directory` is required here `${{ github.workspace }}` is usually assumed by default.

**Comment by Setthawut-Peace** [comment]
Hello, @rdhar  When I try with raw command with
`
terraform init -lock=false -backend-config=env-dev/dev.azurerm.tfbackend
`
`
terraform plan -lock=false -backend-config=env-dev/dev.azurerm.tfbackend
`
It's work

**Comment by Setthawut-Peace** [comment]
More information submodule i have other tf file too like 
.
├── Root/
│   ├── dev/
│   │   ├── dev.tfvars
│   │   └── dev.azurerm.tfbanckend
│   └── Folder_Submodule01/
│       ├── Resource_Group/
│       │   ├── main.tf
│       │   └── variable.tf
│       ├── VNET/
│       │   ├── main.tf
│       │   └── variable.tf
│       └── DB/
│           ├── main.tf
│           └── variable.tf
├── main.tf
├── variable.tf
├── db.tf
├── vnet.tf
└── version.tf

When it enable "fmt" and "validate" I saw they go inside Folder_Submodule01 and fmt in all folder. I think when it run init and plan what folder are there that can make they can't found dev.tfvars and  dev.azurerm.tfbanckend in root

**Comment by rdhar** [comment]
Hi @Setthawut-Peace, TF-via-PR should work the same way as the raw commands. Let's break it up into smaller steps to help troubleshooting.

### ✅ Step 1: Init with raw command
```yaml
- shell: bash
  run: terraform init -lock=false -backend-config=env-dev/dev.azurerm.tfbackend
```

### ⬛  Step 2: Init with hard-coded values
```yaml
- uses: devsectop/tf-via-pr@v12
  with:
    command: init
    arg-lock: false
    arg-backend-config: env-dev/dev.azurerm.tfbackend
```

### ⬛  Step 3: Init + plan with hard-coded values
```yaml
- uses: devsectop/tf-via-pr@v12
  with:
    command: plan
    arg-lock: false
    arg-backend-config: env-dev/dev.azurerm.tfbackend
    arg-var-file: env-dev/dev.tfvars
```

### ⬛  Step 4: Init + plan with variables 
```yaml
- uses: devsectop/tf-via-pr@v12
  with:
    command: plan
    arg-lock: false
    arg-backend-config: env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.azurerm.tfbackend
    arg-var-file: env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.tfvars
```

**Comment by Setthawut-Peace** [comment]
> env-dev/dev.azurerm.tfbackend

@rdhar I change folder and name form **"dev" to "dev-asse"**
###  Step 2: Init with hard-coded values
``` yaml
      - name: Provision TF
        id: provision
        uses: devsectop/tf-via-pr@v12
        with:
          command: init
          arg-backend-config: env-dev-asse/dev-asse.azurerm.tfbackend
          arg-lock: false
```
### Result
```
Error: Failed to read file
The file "'env-dev-asse/dev-asse.azurerm.tfbackend'" could not be read.
Error: Terraform exited with code 1.
Error: Process completed with exit code 1.
```
![Image](https://github.com/user-attachments/assets/297ae2d6-6bf9-4f5e-9150-ac677e4a99f8)




### Step 3: Init + plan with hard-coded values
``` yaml
      - name: Provision TF
        id: provision
        uses: devsectop/tf-via-pr@v12
        with:
          command: plan
          arg-backend-config: env-dev-asse/dev-asse.azurerm.tfbackend
          arg-var-file: env-dev-asse/dev-asse.tfvars
          arg-lock: false
```
### Result
```
Error: Failed to read file

The file "'env-dev-asse/dev-asse.azurerm.tfbackend'" could not be read.
Error: Terraform exited with code 1.
Error: Process completed with exit code 1.
```
![Image](https://github.com/user-attachments/assets/b0111b39-85a4-4cde-a910-c4c4d81e3105)




### Step 4: Init + plan with variables
```yaml
      - name: Provision TF
        id: provision
        uses: devsectop/tf-via-pr@v12
        with:
          command: plan
          arg-backend-config: env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.azurerm.tfbackend
          arg-var-file: env-${{ env.TF_WORKSPACE }}/${{ env.TF_WORKSPACE }}.tfvars
          arg-lock: false
```
### Result
```
Error: Failed to read file
The file "'env-dev-asse/dev-asse.azurerm.tfbackend'" could not be read.
Error: Terraform exited with code 1.
Error: Process completed with exit code 1.
```
![Image](https://github.com/user-attachments/assets/1d7e285a-258d-40aa-b65f-3a4ecac3eda8)

**Comment by rdhar** [comment]
Hi @Setthawut-Peace, what's the reason for changing from "dev" to "dev-asse"? 

I thought you'd said that the raw 'terraform init -lock=false -backend-config=env-dev/dev.azurerm.tfbackend' command worked, so we want to emulate exactly as it is.

**Comment by Setthawut-Peace** [comment]
> Hi [@Setthawut-Peace](https://github.com/Setthawut-Peace), what's the reason for changing from "dev" to "dev-asse"?
> 
> I thought you'd said that the raw 'terraform init -lock=false -backend-config=env-dev/dev.azurerm.tfbackend' command worked, so we want to emulate exactly as it is.

I just change to other season not my Folder and File is update already
![Image](https://github.com/user-attachments/assets/585e5eef-f0f5-4683-9256-8029b231fb18)

**Comment by Setthawut-Peace** [comment]
I try raw command with  `"x"`  `'x'` and `'"x"'` 
I got same error in Action
```
terraform init -lock=false -backend-config="'env-dev-asse/dev-asse.azurerm.tfbackend'"
```
![Image](https://github.com/user-attachments/assets/d58792e4-1c10-48ab-837d-44a9c1b1c8dc)





PASS
```
terraform init -lock=false -backend-config='env-dev-asse/dev-asse.azurerm.tfbackend'
```

```
terraform init -lock=false -backend-config="env-dev-asse/dev-asse.azurerm.tfbackend"
```


If Possible Your Action using `'"x"'` when init or plan and apply ?

**Comment by rdhar** [comment]
Thanks for sharing your breakdown, I think you've cracked it!

The intention for `"'x'"` was to account for path inputs with spaces. For example, `-backend-config="'path to/file.tfbackend'"` but it appears this isn't handled consistent by Terraform CLI.

A new patch should be released within a day to fix, thanks to your contribution!

**Comment by Setthawut-Peace** [comment]
Thank you 

I just try plan and apply too when using terraform plan  `'x'`  will get `Given variables file env-${TF_WORKSPACE}/${TF_WORKSPACE}.tfvars does not exist.`
```
-var-file='env-${TF_WORKSPACE}/${TF_WORKSPACE}.tfvars'
```
![Image](https://github.com/user-attachments/assets/78a7f502-5b08-4ba7-9423-3f7e51ee5146)
![Image](https://github.com/user-attachments/assets/621f35a5-bd76-4bff-af4b-14c267dd4637)

```
terraform plan -var-file="env-${TF_WORKSPACE}/${TF_WORKSPACE}.tfvars"
```
`"x"` It's work

**Comment by rdhar** [comment]
Thanks once again for raising this issue, @Setthawut-Peace! 

This issue has been resolved and you've been credited as a contributor in the latest [v12.0.4](https://github.com/DevSecTop/TF-via-PR/releases/tag/v12.0.4) release.

**Comment by rdhar** [comment]
Happy to announce this is shipped with [v12.0.4](https://github.com/DevSecTop/TF-via-PR/releases/tag/v12.0.4), where you've been credited for your contribution!

Please consider ⭐ this project, if you find it useful.

[![GitHub repository stargazers](https://img.shields.io/github/stars/devsectop/tf-via-pr)](https://github.com/devsectop/tf-via-pr "Become a stargazer.")
